plugin/code_lens.py

```diff
@@ -0,0 +1,185 @@
+from .core.protocol import CodeLens, Range
+from .core.typing import List, Tuple, Dict, Iterable, Generator
+from .core.registry import LspTextCommand
+from .core.registry import windows
+from .core.views import make_command_link
+from .core.views import range_to_region
+from html import escape as html_escape
+import itertools
+import sublime
+
+
+class CodeLensData:
+    __slots__ = (
+        'data',
+        'region',
+        'session_name',
+        'annotation',
+    )
+
+    def __init__(self, data: CodeLens, view: sublime.View, session_name: str) -> None:
+        self.data = data
+        self.region = range_to_region(Range.from_lsp(data['range']), view)
+        self.session_name = session_name
+        self.annotation = '...'
+        self.resolve_annotation()
+
+    def __repr__(self) -> str:
+        return 'CodeLensData(resolved={0}, region={1!r})'.format(self.is_resolved(), self.region)
+
+    def is_resolved(self) -> bool:
+        """A code lens is considered resolved if the inner data contains the 'command' key."""
+        return 'command' in self.data
+
+    def to_lsp(self) -> CodeLens:
+        copy = self.data.copy()
+        copy['session_name'] = self.session_name
+        return copy
+
+    @property
+    def small_html(self) -> str:
+        return '<small style="font-family: system">{}</small>'.format(self.annotation)
+
+    def resolve_annotation(self) -> None:
+        command = self.data.get('command')
+        if command is not None:
+            command_name = command.get('command')
+            if command_name:
+                self.annotation = make_command_link('lsp_execute', command['title'], {
+                    'session_name': self.session_name,
+                    'command_name': command_name,
+                    'command_args': command.get('arguments', []),
+                })
+            else:
+                self.annotation = html_escape(command['title'])
+        else:
+            self.annotation = '...'
+
+    def resolve(self, view: sublime.View, code_lens: CodeLens) -> None:
+        self.data = code_lens
+        self.region = range_to_region(Range.from_lsp(code_lens['range']), view)
+        self.resolve_annotation()
+
+
+class CodeLensView:
+    CODE_LENS_KEY = 'lsp_code_lens'
+
+    def __init__(self, view: sublime.View) -> None:
+        self.view = view
+        self._code_lenses = {}  # type: Dict[Tuple[int, int], List[CodeLensData]]
+
+    def clear(self) -> None:
+        self._code_lenses.clear()
+
+    def is_empty(self) -> bool:
+        return not self._code_lenses
+
+    def clear_annotations(self) -> None:
+        for index, _ in enumerate(self._flat_iteration()):
+            self.view.erase_regions(self._region_key(index))
+
+    def _region_key(self, index: int) -> str:
+        return '{0}.{1}'.format(self.CODE_LENS_KEY, index)
```

Comment:
We're still keying on only the index, not the session. Oh well, maybe for a next step.